url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCServer.java

```diff
@@ -0,0 +1,9048 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ *
+ * The author of this program can be reached at thomas@infolab.northwestern.edu
+ **/
+package soc.server;
+
+import soc.debug.D;  // JM
+
+import soc.game.*;
+import soc.message.*;
+
+import soc.robot.SOCRobotClient;
+import soc.server.database.SOCDBHelper;
+
+import soc.server.genericServer.LocalStringConnection;
+import soc.server.genericServer.Server;
+import soc.server.genericServer.StringConnection;
+
+import soc.util.IntPair;
+import soc.util.SOCGameBoardReset;
+import soc.util.SOCGameList;  // used in javadoc
+import soc.util.SOCRobotParameters;
+import soc.util.Version;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.sql.SQLException;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.Properties;
+import java.util.Random;
+import java.util.StringTokenizer;
+import java.util.Vector;
+
+/**
+ * A server for Settlers of Catan
+ *
+ * @author  Robert S. Thomas
+ *
+ * Note: This is an attempt at being more modular. 5/13/99 RST
+ * Note: Hopefully fixed all of the deadlock problems. 12/27/01 RST
+ *<P>
+ * For server command line options, use the --help option.
+ *<P>
+ * If the database is used (see {@link SOCDBHelper}), users can
+ * be set up with a username & password in that database to log in and play.
+ * Users without accounts can connect by leaving the password blank,
+ * as long as they aren't using a nickname which has a password in the database.
+ *<P>
+ *<b>Network traffic:</b>
+ * The first message over the connection is the client's version
+ * and the second is the server's response:
+ * Either {@link SOCRejectConnection}, or the lists of
+ * channels and games ({@link SOCChannels}, {@link SOCGames}).
+ * See {@link SOCMessage} for details of the client/server protocol.
+ * See {@link Server} for details of the server threading and processing.
+ *<P>
+ * The server supports several <b>debug commands</b> when enabled, and
+ * when sent as chat messages by a user named "debug".
+ * (Or, by the only user in a practice game.)
+ * See {@link #processDebugCommand(StringConnection, String, String)}
+ * for details.
+ *<P>
+ * The version check timer is set in {@link SOCClientData#setVersionTimer(SOCServer, StringConnection)}.
+ * Before 1.1.06, the server's response was first message,
+ * and client version was then sent in reply to server's version.
+ *<P>
+ * Java properties (starting with "jsettlers.") were added in 1.1.09, with constant names
+ * starting with PROP_JSETTLERS_, and listed in {@link #PROPS_LIST}.
+ */
+public class SOCServer extends Server
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This class references 139 other classes. By comparison, 99% of the classes in the CodeGuru reference dataset reference fewer. This indicates that this class is highly coupled with other classes. A class that is highly coupled with other classes is difficult to understand and its behavior might change unexpectedly when one of its referenced classes is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt. We recommend that you simplify this class or break it into multiple classes.